modules/apps/headless/headless-builder/headless-builder-impl/src/main/java/com/liferay/headless/builder/internal/publisher/APIApplicationPublisherImpl.java

```diff
@@ -0,0 +1,113 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.headless.builder.internal.publisher;
+
+import com.liferay.headless.builder.application.APIApplication;
+import com.liferay.headless.builder.application.publisher.APIApplicationPublisher;
+import com.liferay.headless.builder.internal.jaxrs.application.HeadlessBuilderApplication;
+import com.liferay.portal.kernel.util.HashMapDictionaryBuilder;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import javax.ws.rs.core.Application;
+
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.ServiceRegistration;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+
+/**
+ * @author Luis Miguel Barcos
+ */
+@Component(service = APIApplicationPublisher.class)
+public class APIApplicationPublisherImpl implements APIApplicationPublisher {
+
+	@Override
+	public void publish(APIApplication apiApplication) {
+		_headlessBuilderApplicationServiceRegistrationMap.putIfAbsent(
```

Comment:
I think in the invocation to the `putIfAbsent` method, the invocation to `_registerHeadlessBuilderApplication` is performed always, even if the key already exists

Reply:
Done!